src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala

```diff
@@ -27,57 +27,64 @@ object PeerMessageBusActor {
 
   case class MessageFromPeer(message: Message, peerId: PeerId)
 
-  case class Subscription(subscriber: ActorRef, classifier: MessageClassifier)
-
   class PeerMessageBus extends ActorEventBus {
 
     override type Event = MessageFromPeer
     override type Classifier = MessageClassifier
 
-    private var subscriptions: Seq[Subscription] = Nil
+    //FIXME Remove this var
+    private var subscriptions: Map[(Subscriber, PeerSelector), Set[Int]] = Map.empty
 
     override def subscribe(subscriber: ActorRef, to: Classifier): Boolean = {
-      val subscription = Subscription(subscriber, to)
-      if (subscriptions.contains(subscription)) {
-        false
-      } else {
-        subscriptions = subscriptions :+ subscription
+      val newSubscriptions = subscriptions.get((subscriber, to.peerSelector)) match {
+        case Some(messageCodes) =>
+          if (messageCodes == to.messageCodes) subscriptions
```

Comment:
this `if` isn't really needed, since it's a `Set` we can just return `subscriptions + ((subscriber, to.peerSelector) -> (messageCodes ++ to.messageCodes))`, can't we?